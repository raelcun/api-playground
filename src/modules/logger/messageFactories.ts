import { Context } from 'koa'

import { Message } from './types'

export const getMessageFromContext = (ctx: Context): Message => ({
  request: {
    url: ctx.href,
    headers: ctx.headers,
    method: ctx.method,
  },
  response: {
    status: ctx.response.status,
    headers: ctx.response.headers,
    body: ctx.response.body,
  },
})

export const createTimingMessageSegment = (
  startNanos: bigint,
  startTime: Date,
  timingType: string,
): Message['timing'] => {
  const duration = Number(process.hrtime.bigint() - startNanos) / 1e6
  const startTimestamp = startTime.toISOString()
  const endTimestamp = new Date(startTime.getTime() + duration).toISOString()

  return {
    type: timingType,
    start: startTimestamp,
    end: endTimestamp,
    duration: Number(process.hrtime.bigint() - startNanos) / 1e6,
  }
}
