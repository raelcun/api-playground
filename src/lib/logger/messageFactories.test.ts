import { Context } from 'koa'

import { createMockContext } from '@shopify/jest-koa-mocks'

import { createTimingMessageSegment, getMessageFromContext } from './messageFactories'

const createContext = (): Context => {
  const requestBody = { foo: 'bar' }
  const responseBody = { fizz: 'buzz' }
  const context: Context = createMockContext({
    requestBody,
    headers: {
      foo: 'bar',
    },
  })
  context.response.status = 200
  context.response.body = responseBody
  context.set('fizz', 'buzz')

  return context
}

describe('log message factories', () => {
  describe('getMessageFromContext', () => {
    test.each([
      ['when options are defaulted', undefined],
      ['when request options is true', { request: true, response: false }],
      ['when response options is true', { request: false, response: true }],
      ['when request and response is false', { request: false, response: false }],
      ['when request and response is true', { request: true, response: true }],
    ])('should set message properties %s', (_, options) => {
      expect(getMessageFromContext(createContext(), options)).toMatchSnapshot()
    })
  })

  test('createTimingMessageSegment', () => {
    const time = new Date()
    const message = createTimingMessageSegment(process.hrtime.bigint(), time, 'MOCK_TIMING')
    expect(message.type).toEqual('MOCK_TIMING')
    expect(message.start).toEqual(time.toISOString())
    expect(time.getTime() - Date.parse(message.end)).toBeGreaterThanOrEqual(0)
    expect(time.getTime() - Date.parse(message.end)).toBeLessThan(20)
    expect(message.duration).toBeGreaterThan(0)
  })
})
