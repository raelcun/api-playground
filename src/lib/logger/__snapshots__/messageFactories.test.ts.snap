// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`log message factories getMessageFromContext should set message properties when options are defaulted 1`] = `
Object {
  "request": Object {
    "headers": Object {
      "foo": "bar",
      "host": "test.com",
    },
    "method": "GET",
    "url": "http://test.com/",
  },
  "response": Object {
    "body": Object {
      "fizz": "buzz",
    },
    "headers": Object {
      "content-type": "application/json; charset=utf-8",
      "fizz": "buzz",
    },
    "status": 200,
  },
}
`;

exports[`log message factories getMessageFromContext should set message properties when request and response is false 1`] = `Object {}`;

exports[`log message factories getMessageFromContext should set message properties when request and response is true 1`] = `
Object {
  "request": Object {
    "headers": Object {
      "foo": "bar",
      "host": "test.com",
    },
    "method": "GET",
    "url": "http://test.com/",
  },
  "response": Object {
    "body": Object {
      "fizz": "buzz",
    },
    "headers": Object {
      "content-type": "application/json; charset=utf-8",
      "fizz": "buzz",
    },
    "status": 200,
  },
}
`;

exports[`log message factories getMessageFromContext should set message properties when request options is true 1`] = `
Object {
  "request": Object {
    "headers": Object {
      "foo": "bar",
      "host": "test.com",
    },
    "method": "GET",
    "url": "http://test.com/",
  },
}
`;

exports[`log message factories getMessageFromContext should set message properties when response options is true 1`] = `
Object {
  "response": Object {
    "body": Object {
      "fizz": "buzz",
    },
    "headers": Object {
      "content-type": "application/json; charset=utf-8",
      "fizz": "buzz",
    },
    "status": 200,
  },
}
`;
